src/realm/table.cpp

```diff
@@ -1080,70 +1093,117 @@ void Table::set_table_type(Type table_type)
+        throw std::logic_error(
+            util::format("Cannot change '%1' to embedded when using a primary key.", get_class_name()));
     }
 
-    // `has_backlink_columns` indicates if the table is embedded in any other table.
-    bool has_backlink_columns = false;
-    for_each_backlink_column([&has_backlink_columns](ColKey) {
-        has_backlink_columns = true;
-        return IteratorControl::Stop;
-    });
-    if (!has_backlink_columns) {
-        throw std::logic_error(
-            util::format("Cannot change '%1' to embedded without backlink columns. Objects must be embedded in "
-                         "at least one other class.",
-                         get_name()));
-    }
-    else if (size() > 0) {
-        for (auto object : *this) {
-            size_t backlink_count = 0;
-            for_each_backlink_column([&](ColKey backlink_col_key) {
-                size_t cur_backlinks = object.get_backlink_cnt(backlink_col_key);
-                if (cur_backlinks > 0) {
-                    // Make sure this link is not an untyped ObjLink which lacks core support in many places
-                    ColKey source_col = get_opposite_column(backlink_col_key);
-                    REALM_ASSERT(source_col); // backlink columns should always have a source
-                    TableRef source_table = get_opposite_table(backlink_col_key);
-                    ColKey forward_col_mapped = source_table->get_opposite_column(source_col);
-                    if (!forward_col_mapped) {
-                        throw std::logic_error(util::format("There is a dynamic/untyped link from a Mixed property "
-                                                            "'%1.%2' which prevents migrating class '%3' to embedded",
-                                                            source_table->get_name(),
-                                                            source_table->get_column_name(source_col), get_name()));
+    if (size() == 0) {
+        do_set_table_type(Type::Embedded);
+        return;
+    }
+
+    // Check all of the objects for invalid incoming links. Each embedded object
+    // must have exactly one incoming link, and it must be from a non-Mixed property.
+    // Objects with no incoming links are either deleted or an error (depending
+    // on `handle_backlinks`), and objects with multiple incoming links are either
+    // cloned for each of the incoming links or an error (again depending on `handle_backlinks`).
+    // Incoming links from a Mixed property are always an error, as those can't
+    // link to embedded objects
+    ArrayInteger leaf(get_alloc());
+    enum class LinkCount : int8_t { None, One, Multiple };
+    std::vector<LinkCount> incoming_link_count;
+    std::vector<ObjKey> orphans;
+    std::vector<ObjKey> multiple_incoming_links;
+    traverse_clusters([&](const Cluster* cluster) {
```

Comment:
This was probably not actually a useful thing to optimize, but processing each backlink array leaf in order rather than processing each obj in order is a pretty dramatic speedup.